src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs

```diff
@@ -0,0 +1,175 @@
+        {
+            FundingSource.Levy,
+            FundingSource.CoInvestedEmployer,
+            FundingSource.CoInvestedSfa,
+            FundingSource.FullyFundedSfa
+        };
+
+        public LearnerRefundProcessor(ILogger logger)
+        {
+            _logger = logger;
+        }
+
+        public List<PaymentEntity> ProcessRefundsForLearner(
+            List<RequiredPaymentEntity> refunds,
+            List<HistoricalPaymentEntity> previousPayments)
+        {
+            var refundGroups = refunds.ToLookup(x => new RefundGroup(x));
+            var volatilePreviousPayments = new List<HistoricalPaymentEntity>(previousPayments);
+
+            var refundPayments = new List<PaymentEntity>();
+            
+            foreach (var refundGroup in refundGroups)
+            {
+                foreach (var refund in refundGroup)
+                {
+                    var previousPaymentGroup = volatilePreviousPayments.ToLookup(x => new RefundGroup(x));
+                    var previousPaymentsForRefundGroup = new List<HistoricalPaymentEntity>();
+                    try
+                    {
+                        if (previousPaymentGroup.Contains(refundGroup.Key))
+                        {
+                            previousPaymentsForRefundGroup = previousPaymentGroup[refundGroup.Key].ToList();
+                        }
+
+                        var newRefunds = ProcessRefund(refund, previousPaymentsForRefundGroup);
+                        refundPayments.AddRange(newRefunds);
+                        volatilePreviousPayments.AddRange(newRefunds.Select(x => new HistoricalPaymentEntity(x, refund)));
+                    }
+                    catch(ApplicationException)
+                    { 
+                        // A funding source is negative, so ignoring this refund
+                        _logger.Error($"Refund for UKPRN: {refund.Ukprn} " +
+                                      $"LearnRefNumber: {refund.LearnRefNumber} " +
+                                      $"Amount: {refund.AmountDue} " +
+                                      $"DeliveryMonth: {refund.DeliveryMonth} " +
+                                      $"not made due to a net negative funding source");
+                    }
+                }
+            }
+
+            return refundPayments;
+        }
+
+        private List<PaymentEntity> ProcessRefund(
+            RequiredPaymentEntity refund, 
+            List<HistoricalPaymentEntity> previousPayments)
+        {
+            var refundPayments = new List<PaymentEntity>();
+
+            var amountToRefund = refund.AmountDue;
+            var amountRefunded = 0m;
+            var month = refund.DeliveryMonth;
+            var year = refund.DeliveryYear;
+
+            while (amountToRefund < amountRefunded)
+            {
+                var stillToRefund = amountToRefund - amountRefunded;
+                var paymentsForPeriod = previousPayments
```

Comment:
This variable paymentforPeriod is only used in the ProcessRefundForPeriod() it should exist there as you also are passing in the filter values